File: Leetcode/0025_reverse_nodes_in_k_groups/0025_reverse_nodes_in_k_groups.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

 type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	reverse := func(start, end *ListNode) *ListNode {
		var prev *ListNode
		current := start
		for current != end {
			nextNode := current.Next
			current.Next = prev
			prev = current
			current = nextNode
		}
		return prev
	}

	count := 0
	current := head
	for current != nil && count < k {
		current = current.Next
		count++
	}

	if count < k {
		return head
	}

	reversedHead := reverse(head, current)
	if head != nil {
		head.Next = reverseKGroup(current, k)
	}

	return reversedHead
}

File: leetcode/0025_reverse_nodes_in_k_groups/0025_reverse_nodes_in_k_groups.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class ListNode {
    val: number;
    next: ListNode | null;
    
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}

function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
    function reverse(start: ListNode | null, end: ListNode | null): ListNode | null {
        let prev: ListNode | null = null;
        let current: ListNode | null = start;
        while (current !== end) {
            const nextNode: ListNode | null = current!.next;
            current!.next = prev;
            prev = current;
            current = nextNode;
        }
        return prev;
    }

    let count: number = 0;
    let current: ListNode | null = head;
    while (current !== null && count < k) {
        current = current.next;
        count++;
    }

    if (count < k) {
        return head;
    }

    const reversedHead: ListNode | null = reverse(head, current);
    if (head !== null) {
        head.next = reverseKGroup(current, k);
    }

    return reversedHead;
}
